ros/src/tl_detector/light_classification/tl_classifier.py

```diff
@@ -1,9 +1,206 @@
+
+    model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
+
+    train_samples, valid_samples = create_data()
+    print('{} training, {} validation samples'.format(train_samples, valid_samples))
+
+    train_datagen = ImageDataGenerator(
+        preprocessing_function=preprocess_input,
+        rotation_range=5,
+        width_shift_range=.05,
+        height_shift_range=.05,
+        zoom_range=.01,
+        horizontal_flip=True)
+    test_datagen = ImageDataGenerator(
+        preprocessing_function=preprocess_input
+    )
+
+    batch_size = 8
+    train_generator = train_datagen.flow_from_directory(
+        'src/tl_detector/light_classification/train',
+        target_size=(IMG_WIDTH, IMG_HEIGHT),
+        batch_size=batch_size,
+        class_mode='categorical')
+
+    validation_generator = test_datagen.flow_from_directory(
+        'src/tl_detector/light_classification/valid',
+        target_size=(IMG_WIDTH, IMG_HEIGHT),
+        batch_size=batch_size,
+        class_mode='categorical')
+
+    model.fit_generator(
+        train_generator,
+        steps_per_epoch=int(1.5*train_samples/batch_size),
+        epochs=10,
+        validation_data=validation_generator,
+        validation_steps=int(valid_samples/batch_size),
+        callbacks=[
+            ModelCheckpoint(
+                'src/tl_detector/model.h5',
+                verbose=2,
+                save_weights_only=False,
+                save_best_only=True
+            )
+        ])
+
+def train_site():
+    model = create_model(4)
+    model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])
+    train_datagen = ImageDataGenerator(
+        preprocessing_function=preprocess_input,
+        rotation_range=5,
+        width_shift_range=.1,
+        height_shift_range=.1,
+        zoom_range=.05,
+        horizontal_flip=True)
+
+    batch_size = 16
+    train_generator = train_datagen.flow_from_directory(
+            'src/tl_detector/light_classification/site_training_data',
+            target_size=(IMG_WIDTH, IMG_HEIGHT),
+            batch_size=batch_size,
+#            save_to_dir='src/tl_detector/light_classification/aug',
+            class_mode='categorical')
+
+    train_samples = 191
+
+    model.fit_generator(
+            train_generator,
+            steps_per_epoch=int(train_samples/batch_size),
+            epochs=20,
+            validation_data=train_generator,
+            validation_steps=int(train_samples/batch_size),
+            callbacks=[
+                ModelCheckpoint(
+                    'src/tl_detector/site_model.h5',
+                    verbose=2,
+                    save_weights_only=False,
+                    save_best_only=True
+                )
+            ])
+
+
 class TLClassifier(object):
     def __init__(self):
-        #TODO load classifier
-        pass
+        self.model = None
+
+        if rospy.get_param('/launch') == 'styx':
```

Comment:
I wonder if the site uses the same launch script? If not `get_param` might raise a`KeyError` if `/launch` is not found. Maybe it's better to default to site unless the parameter is available and says `styx`

Reply:
I defined the parameters in both `launch` files so this should work!

Reply:
I was worried that they will swap out the launch script for site. Maybe Carla has other ROS nodes that's not included in the launch script here. Anyway I probably worried too much.

Reply:
That makes sense—I think the key error would be good in that case, kind of acts like a compilation error...